mdf_matio/adapters/generic.py

```diff
@@ -0,0 +1,68 @@
+import os
+
+import jsonschema
+import mdf_toolbox
+
+from materials_io.adapters.base import BaseAdapter
+
+
+class GenericMDFAdapter(BaseAdapter):
+    """Generic adapter for MDF extractors. Adapts metadata with MDF-format fields present."""
+
+    def __init__(self, schema_branch=None, schema_uri=None):
+        """Create a GenericMDFAdapter object.
+
+        Arguments:
+            schema_branch (str): The branch of the MDF data-schemas Github repository to
+                    download and validate against. Default None, to instead use local files.
+            schema_uri (str): The uri to the MDF schema location. Local and nonlocal locations
+                    are supported. Default None, to use the Github branch instead
+
+        Note:
+            Exactly one of schema_branch and schema_path must be supplied.
+        """
+        # Check that schema_branch XOR schema_uri is True
+        if not bool(schema_branch) ^ bool(schema_uri):
+            raise ValueError("Exactly one of schema_branch and schema_uri must be supplied.")
+
+        # Generate URI from Github if branch supplied
+        if schema_branch:
+            schema_uri = ("https://raw.githubusercontent.com/materials-data-facility/"
+                          "data-schemas/{}/schemas/".format(schema_branch))
+        # If URI is bare file path, make into URI
+        elif os.path.exists(schema_uri):
+            schema_uri = "{}{}{}".format(
+                                    "file://" if not schema_uri.startswith("file://") else "",
+                                    os.path.abspath(schema_uri),
+                                    "/" if schema_uri.endswith("/") else "")
+        # Fetch record schema with resolver
+        resolver = jsonschema.RefResolver(schema_uri, None)
```

Comment:
I'd recommend creating the RefResolver as a module-level global variable. Otherwise, each parser has their own RefResolver and you will not benefit from the caching logic

Reply:
That's how I approached it in the `validation.py` module

Reply:
The user must be able to set the schema to validate against (e.g. `dev` or `master` branch). The earliest they can reasonably do that is in the `__init__` call for an object, because they should be able to import this class once, but validate against multiple schemas (with a separate instance for each schema). I'm not seeing any way to make the `RefResolver` a global variable without breaking some of this functionality.

Reply:
Yea, let's make it a class attribute then during the `__init__` call then.